salt/_modules/metalk8s.py

```diff
@@ -29,3 +44,42 @@ def get_etcd_endpoint():
         return '{host}=https://{ip}:{port}'.format(
             host=hostname, ip=ips[0], port=DEFAULT_ETCD_PORT
         )
+
+
+def add_etcd_node(host, endpoint):
+    '''Add a new `etcd` node into the `etcd` cluster.
+
+    Arguments:
+        input_files (str): hostname of the new etcd node
+        output_file (str): endpoint of the new etcd node
+    '''
+    pod_name = 'etcd-{}'.format(__salt__['network.get_hostname']())
+    try:
+        client = kubernetes.config.new_client_from_config(
+            config_file='/etc/kubernetes/admin.conf'
```

Comment:
Yup, indeed it's worth to mention it.

Reply:
If it's only to run on the `salt master` you should use a runner not a module. 
Here it's works because we use an orchestrate so we can run "modules" on the master but normally it's only runner who run on the master.

Reply:
So where should I put this code?

Reply:
Since it's a runner in `salt/_runners/xxx.py`

Reply:
I disagree, @TeddyAndrieux. This is an action that *could* be performed on any node (and, indeed, at some point we may decide to actually run it on a non-master minion, orchestrated by a runner). As such, it's a module. We simply happen to invoke it on the master for now.

Here's how I understand the Salt taxonomy:

- Module procedure: and 'action', something to be executed somewhere (likely a minion), very imperative.
- State procedure: code which exposes functionality to idempotent'ly reach a target state on a system by invoking 'module' actions. Imperative, in a way. In Ansible terms: task implementation I believe.
- State SLS: combination/invocation of state procedures making up a final state. Very declarative. In Ansible terms: role I believe.

- Runner procedure: invoked in context of the master, can invoke module procedures on master or minions, and other runners (again invoked in master context). Used to orchestrate things, mainly. In Ansible terms: playbook I believe.
- `state.orchestrate` runner: special case of a `runner` procedure which uses an SLS as 'input' and interprets it as a DSL instead of requiring one to write the Python code of the same effect.

Reply:
I agree but for me the goal was that the master is the only "salt" who talk to the apiserver, No ?

Reply:
- Currently this code goes through the API-server because it uses an `exec` in an `etcd` Pod to perform a certain action. That's not the right approach, cfr. #878.
- Even when #878 gets implemented, the call would still happen from the master because that's the only one who's supposed to have the right credentials to invoke the command.
- This is a step which should be executed 'once', so it's part of 'orchestration'.

However.

Whilst we may invoke this procedure on the master, and as part of a larger playbook/orchestration, that doesn't mean the functionality itself (i.e. the procedure) couldn't be useful in other contexts. So, it's a module procedure, which may be invoked from an orchestration/runner, or in a state module if ever, or ...

In general:

- An 'action', imperative procedure to affect 'real world': module procedure.
- An 'orchestration', steps to be followed with centralized command (~ running on the master): runner procedure (or `state.orchestrate` SLS, which is kinda the same).

The question of 'where do *we* currently execute this code' is not part of the equation.

Reply:
Ok make sense